File: src/go/ch01/1.2/l2/karatsuba.go
package main

// Karatsuba multiplies two non-negative integers using the Karatsuba method.
// Falls back to builtin for small values. Works for int range; for very large
// integers, prefer math/big (not covered here to keep the example focused).
func Karatsuba(x, y int) int {
	if x < 10 || y < 10 {
		return x * y
	}
	// choose split position by bit-length
	n := max(bitLen(x), bitLen(y))
	m := n / 2

	high1, low1 := x>>m, x-(x>>m)<<m
	high2, low2 := y>>m, y-(y>>m)<<m

	z0 := Karatsuba(low1, low2)
	z2 := Karatsuba(high1, high2)
	z1 := Karatsuba(low1+high1, low2+high2) - z0 - z2

	return (z2 << (2 * m)) + (z1 << m) + z0
}

func bitLen(v int) int {
	if v == 0 {
		return 0
	}
	l := 0
	for vv := v; vv > 0; vv >>= 1 {
		l++
	}
	return l
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

File: archived/v1/src/erlang/ch01/1.2/src/karatsuba.erl
-module(karatsuba).
-export([mul/2, demo/0]).

%% Karatsuba multiplication for non-negative integers
mul(X, Y) when X < 10; Y < 10 ->
    X * Y;
mul(X, Y) ->
    N = max(bit_size(X), bit_size(Y)),
    M = N div 2,
    High1 = X bsr M,
    Low1  = X - (High1 bsl M),
    High2 = Y bsr M,
    Low2  = Y - (High2 bsl M),
    Z0 = mul(Low1, Low2),
    Z2 = mul(High1, High2),
    Z1 = mul(Low1+High1, Low2+High2) - Z0 - Z2,
    (Z2 bsl (2*M)) + (Z1 bsl M) + Z0.

bit_size(0) -> 0;
bit_size(N) -> bit_size(N, 0).
bit_size(0, Acc) -> Acc;
bit_size(N, Acc) -> bit_size(N bsr 1, Acc+1).

demo() ->
    A = 1234, B = 5678,
    Res = mul(A,B),
    io:format("Karatsuba result: ~p~n", [Res]),
    io:format("Expected: ~p~n", [A*B]).
